File: src/Chapter08/Part07/AbstractMachine.java
package Part07;

import java.util.LinkedList;
import java.util.List;

public class AbstractMachine {
    /* EXPRESSIONS */
    sealed interface Expr permits Val, Add {
    }

    record Val(int value) implements Expr {
    }

    record Add(Expr left, Expr right) implements Expr {
    }

    /* OPERATIONS */
    sealed interface Op permits EVAL, ADD {
    }

    record EVAL(Expr expr) implements Op {
    }

    record ADD(int value) implements Op {
    }

    static int eval(Expr expr, List<Op> cont) {
        return switch (expr) {
            case Val n -> exec(n.value, cont);
            case Add e -> {
                cont.addFirst(new EVAL(e.right));
                yield eval(e.left, cont);
            }
        };
    }

    static int exec(int n, List<Op> cont) {
        if (cont.isEmpty())
            return n;

        return switch (cont.removeFirst()) {
            case EVAL op -> {
                cont.addFirst(new ADD(n));
                yield eval(op.expr, cont);
            }
            case ADD op -> exec(n + op.value, cont);
        };
    }

    static int value(Expr expr) {
        return eval(expr, new LinkedList<>());
    }

    public static void main(String[] args) {
        Expr expr = new Add(
                new Add(
                        new Val(2),
                        new Val(3)),
                new Val(4));

        int result = value(expr);

        assert (result == 9); // Add the `-ea` JVM option to activate assertions
        System.out.println("Result: " + result);
    }
}

File: src/Chapter08/Part07/abstract_machine.ml
(*

cd ./src/Chapter08/Part07/
dune exec ./abstract_machine.exe

*)

type op =
  | EVAL of expr
  | ADD of int

and expr =
  | Val of int
  | Add of expr * expr

let rec eval expr cont =
  match expr with
  | Val n -> exec cont n
  | Add (x, y) -> eval x (EVAL y :: cont)

and exec cont n =
  match cont with
  | [] -> n
  | EVAL y :: rest -> eval y (ADD n :: rest)
  | ADD m :: rest -> exec rest (n + m)

let value e = eval e []

let () =
  let
    expr = Add
            ( Add
                ( Val 2
                , Val 3
                )
            , Val 4
            )
  in
  let result = value expr in
  assert(9 = result);
  Printf.printf "%d\n" result
